File: app/src/main/java/com/daniel/mobilepauker2/PaukerManager.java
/*
 * Copyright 2011 Brian Ford
 *
 * This file is part of Pocket Pauker.
 *
 * Pocket Pauker is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * Pocket Pauker is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * See http://www.gnu.org/licenses/.

 */

package com.daniel.mobilepauker2;

import android.app.Activity;
import android.content.Context;
import android.os.Environment;
import android.widget.Toast;

import com.daniel.mobilepauker2.model.ModelManager;
import com.daniel.mobilepauker2.model.pauker_native.Lesson;
import com.daniel.mobilepauker2.model.xmlsupport.FlashCardXMLPullFeedParser;
import com.daniel.mobilepauker2.utils.Constants;
import com.daniel.mobilepauker2.utils.Log;

import java.io.File;
import java.io.FileFilter;
import java.io.IOException;
import java.net.URI;

public class PaukerManager {

    private static PaukerManager instance = null;

    private String mCurrentFileName = Constants.DEFAULT_FILE_NAME;
    private String mFileAbsolutePath = null;
    private boolean mSaveRequired = false;

    private PaukerManager() {
    }

    public static PaukerManager instance() {
        if (instance == null) {
            instance = new PaukerManager();
        }
        return instance;
    }

    // TODO Toaster
    public static void showToast(final Activity context, final String text, final int duration) {
        context.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if (text != null && !text.isEmpty()) {
                    Toast.makeText(context, text, duration).show();
                }
            }
        });
    }

    // TODO Toaster
    public static void showToast(Activity context, int textResource, int duration) {
        showToast(context, context.getString(textResource), duration);
    }

    /**
     * Get the default application data directory.
     * <p>
     * Note that this is not necessarily where the current file has been loaded from.
     * @return Default application data directory
     */
    public String getApplicationDataDirectory() {
        // The application data directory can change when a file is loaded
        return Constants.DEFAULT_APP_FILE_DIRECTORY;
    }

    /**
     * Setup a new lesson in the default application directory
     * <p>
     * Note the lesson is not created until it is saved
     * Note this method appends .pau.gz to the application filename.
     */
    public void setupNewApplicationLesson() { // TODO LessonManager
        /*String _filename = filename + ".pau.gz";
        if (setCurrentFileName(_filename)) {
            mApplicationDataDirectory = Constants.DEFAULT_APP_FILE_DIRECTORY;
            String filePath = Environment.getExternalStorageDirectory() + getApplicationDataDirectory() + _filename;
            File file = new File(filePath);
            setCurrentFileName(file.getName());
            setFileAbsolutePath(file.getAbsolutePath());
            ModelManager.instance().createNewLesson(filename);
            return true;
        } else {
            return false;
        }*/
        mCurrentFileName = Constants.DEFAULT_FILE_NAME;
        ModelManager.instance().createNewLesson();
    }

    public String getFileAbsolutePath() {
        return mFileAbsolutePath;
    }

    private void setFileAbsolutePath(String fileAbsolutePath) {
        // Validate the filename
        if (fileAbsolutePath == null) {
            return;
        }

        mFileAbsolutePath = fileAbsolutePath;
    }

    public boolean setCurrentFileName(String filename) {
        if (!filename.endsWith(".pau.gz"))
            filename += ".pau.gz";

        // Validate the filename
        if (!isNameValid(filename)) return false;

        mCurrentFileName = filename;
        return true;
    }

    public boolean isNameValid(String filename) {
        if (filename == null || filename.isEmpty() || isNameEmpty(filename)) {
            return false;
        }
        return validateFileEnding(filename);
    }

    public boolean isNameEmpty(String fileName) {
        for (String ending : Constants.PAUKER_FILE_ENDING) {
            if (fileName.equals(ending)) return true;
        }
        return false;
    }

    public String getCurrentFileName() {
        return mCurrentFileName;
    }

    public boolean isSaveRequired() {
        return mSaveRequired;
    }

    public void setSaveRequired(boolean saveRequired) {
        mSaveRequired = saveRequired;
    }

    /**
     * Gibt den Namen der aktuellen Lektion ohne Endungen zurück.
     * @return Lektionsname ohne Endungen
     */
    public String getReadableFileName() {
        String filename = mCurrentFileName;

        return getReadableFileName(filename);
    }

    /**
     * Loads a lesson from a file
     * @param filename Name der Datei, die importiert werden soll
     * @return <b>True</b> if lesson loaded ok
     */
    public File getFilePath(Context context, String filename) throws IOException {
        // Validate the filename
        if (!validateFilename(filename)) {
            PaukerManager.showToast((Activity) context, R.string.error_filename_invalid, Toast.LENGTH_LONG);
            throw new IOException("Filename invalid");
        }

        String filePath = Environment.getExternalStorageDirectory() + getApplicationDataDirectory() + filename;
        return new File(filePath);
    }

    /**
     * Gibt den Namen der Lektion ohne Endungen zurück.
     * @param filename Lektionsname
     * @return Lektionsname ohne Endungen
     */
    public String getReadableFileName(String filename) {
        if (validateFileEnding(filename)) {
            return filename.substring(0, filename.length() - 7);
        } else if (filename.endsWith(".pau") || filename.endsWith(".xml")) {
            return filename.substring(0, filename.length() - 4);
        } else {
            return filename;
        }
    }

    public boolean validateFilename(String filename) {
        if (filename == null) {
            Log.d("Validate Filename", "File name is invalid");
            return false;
        }

        if (!validateFileEnding(filename)) {
            Log.d("Validate Filename", "File not ending with .pau.gz");
            return false;
        }

        return true;
    }

    private boolean validateFileEnding(String fileName) {
        for (String ending : Constants.PAUKER_FILE_ENDING) {
            if (fileName.endsWith(ending)) {
                return true;
            }
        }
        return false;
    }

    public File[] listFiles(final Context context) throws SecurityException {
        File appDirectory = new File(Environment.getExternalStorageDirectory() + getApplicationDataDirectory());
        File[] files = new File[0];

        if (!appDirectory.exists() && !appDirectory.mkdir()) {
            return files;
        }

        if (appDirectory.exists() && appDirectory.isDirectory()) {
            files = appDirectory.listFiles(new FileFilter() {
                @Override
                public boolean accept(File file) {
                    return validateFilename(file.getName());
                }
            });
        } else {
            showToast((Activity) context, R.string.error_importflashcardfile_directory, Toast.LENGTH_LONG);
            return files;
        }
        return files;
    }

    public boolean isFileExisting(Context context, String fileName) {
        File[] files = listFiles(context);

        for (File file : files) {
            if (file.getName().equals(fileName)) {
                return true;
            }
        }
        return false;
    }

    public void loadLessonFromFile(File file) throws IOException {
        URI uri = file.toURI();
        FlashCardXMLPullFeedParser xmlFlashCardFeedParser = new FlashCardXMLPullFeedParser(uri.toURL());
        Lesson lesson = xmlFlashCardFeedParser.parse();
        setCurrentFileName(file.getName());
        setFileAbsolutePath(file.getAbsolutePath());
        ModelManager.instance().setLesson(lesson);
    }
}

File: app/src/main/java/com/daniel/mobilepauker2/core/PaukerManager.kt
/*
 * Copyright 2011 Brian Ford
 *
 * This file is part of Pocket Pauker.
 *
 * Pocket Pauker is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * Pocket Pauker is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * See http://www.gnu.org/licenses/.

 */
package com.daniel.mobilepauker2.core

import android.app.Activity
import android.content.Context
import android.os.Environment
import android.widget.Toast
import com.daniel.mobilepauker2.R
import com.daniel.mobilepauker2.pauker_native.ModelManager
import com.daniel.mobilepauker2.core.model.xmlsupport.FlashCardXMLPullFeedParser
import com.daniel.mobilepauker2.pauker_native.Log
import java.io.File
import java.io.IOException

class PaukerManager(val modelManager: ModelManager) {

    var currentFileName: String? =
            Constants.DEFAULT_FILE_NAME
        private set
    private var mFileAbsolutePath: String? = null
    var isSaveRequired = false

    /**
     * Get the default application data directory.
     *
     *
     * Note that this is not necessarily where the current file has been loaded from.
     * @return Default application data directory
     */
    val applicationDataDirectory: String?
        get() =// The application data directory can change when a file is loaded
            Constants.DEFAULT_APP_FILE_DIRECTORY

    /**
     * Setup a new lesson in the default application directory
     *
     *
     * Note the lesson is not created until it is saved
     * Note this method appends .pau.gz to the application filename.
     */
    fun setupNewApplicationLesson() { /*String _filename = filename + ".pau.gz";
        if (setCurrentFileName(_filename)) {
            mApplicationDataDirectory = Constants.DEFAULT_APP_FILE_DIRECTORY;
            String filePath = Environment.getExternalStorageDirectory() + getApplicationDataDirectory() + _filename;
            File file = new File(filePath);
            setCurrentFileName(file.getName());
            setFileAbsolutePath(file.getAbsolutePath());
            ModelManager.instance().createNewLesson(filename);
            return true;
        } else {
            return false;
        }*/
        currentFileName = Constants.DEFAULT_FILE_NAME
        modelManager.createNewLesson()
    }

    // Todo replace this with the File Class
    var fileAbsolutePath: String?
        get() = mFileAbsolutePath
        private set(fileAbsolutePath) { // Validate the filename
            if (fileAbsolutePath == null) {
                return
            }
            mFileAbsolutePath = fileAbsolutePath
        }

    fun setCurrentFileName(filename: String): Boolean {
        var newFilename = filename
        if (!newFilename.endsWith(".pau.gz")) newFilename += ".pau.gz"
        // Validate the filename
        if (!isNameValid(newFilename)) return false
        currentFileName = newFilename
        return true
    }

    fun isNameValid(filename: String?): Boolean {
        return if (filename == null || filename.isEmpty() || isNameEmpty(filename)) {
            false
        } else validateFileEnding(filename)
    }

    fun isNameEmpty(fileName: String): Boolean {
        for (ending in Constants.PAUKER_FILE_ENDING) {
            if (fileName == ending) return true
        }
        return false
    }

    /**
     * Gibt den Namen der aktuellen Lektion ohne Endungen zurück.
     * @return Lektionsname ohne Endungen
     */
    val readableFileName: String?
        get() {
            val filename = currentFileName
            return getReadableFileName(filename)
        }

    /**
     * Loads a lesson from a file
     * @param filename Name der Datei, die importiert werden soll
     * @return **True** if lesson loaded ok
     */
    @Throws(IOException::class)
    fun getFilePath(
            context: Context,
            filename: String?
    ): File { // Validate the filename
        if (!validateFilename(filename)) {
            showToast(
                    context as Activity,
                    R.string.error_filename_invalid,
                    Toast.LENGTH_LONG
            )
            throw IOException("Filename invalid")
        }
        val filePath =
                Environment.getExternalStorageDirectory().toString() + applicationDataDirectory + filename
        return File(filePath)
    }

    /**
     * Gibt den Namen der Lektion ohne Endungen zurück.
     * @param filename Lektionsname
     * @return Lektionsname ohne Endungen
     */
    fun getReadableFileName(filename: String?): String? {
        return if (validateFileEnding(filename)) {
            filename!!.substring(0, filename.length - 7)
        } else if (filename!!.endsWith(".pau") || filename.endsWith(".xml")) {
            filename.substring(0, filename.length - 4)
        } else {
            filename
        }
    }

    fun validateFilename(filename: String?): Boolean {
        if (filename == null) {
            Log.d("Validate Filename", "File name is invalid")
            return false
        }
        if (!validateFileEnding(filename)) {
            Log.d(
                    "Validate Filename",
                    "File not ending with .pau.gz"
            )
            return false
        }
        return true
    }

    private fun validateFileEnding(fileName: String?): Boolean {
        for (ending in Constants.PAUKER_FILE_ENDING) {
            if (fileName!!.endsWith(ending)) {
                return true
            }
        }
        return false
    }

    @Throws(SecurityException::class)
    fun listFiles(context: Context): Array<File>? {
        val appDirectory =
                File(Environment.getExternalStorageDirectory().toString() + applicationDataDirectory)
        val files: Array<File>
        if (!appDirectory.exists() && !appDirectory.mkdir()) {
            throw SecurityException()
        }
        files = if (appDirectory.exists() && appDirectory.isDirectory) {
            appDirectory.listFiles { file -> validateFilename(file.name) }
        } else {
            showToast(
                    context as Activity,
                    R.string.error_importflashcardfile_directory,
                    Toast.LENGTH_LONG
            )
            return null
        }
        return files
    }

    fun isFileExisting(context: Context, fileName: String): Boolean {
        try {
            val files = listFiles(context)
            for (file in files!!) {
                if (file.name == fileName) {
                    return true
                }
            }
        } catch (ignored: SecurityException) {
        }
        return false
    }

    @Throws(IOException::class)
    fun loadLessonFromFile(file: File?) {
        val uri = file!!.toURI()
        val xmlFlashCardFeedParser = FlashCardXMLPullFeedParser(uri.toURL())
        val lesson = xmlFlashCardFeedParser.parse()
        setCurrentFileName(file.name)
        fileAbsolutePath = file.absolutePath
        modelManager.lesson = lesson
    }

    companion object {
        lateinit var manager: PaukerManager
        fun instance(): PaukerManager {
            return manager
        }

        fun showToast(context: Activity, text: String?, duration: Int) {
            context.runOnUiThread {
                if (text != null && text.isNotEmpty()) {
                    Toast.makeText(context, text, duration).show()
                }
            }
        }

        fun showToast(context: Activity, textResource: Int, duration: Int) {
            showToast(context, context.getString(textResource), duration)
        }
    }
}